File: TodoList/app/src/main/java/com/callor/todo/FirstFragment.java
package com.callor.todo;

import android.os.Bundle;

import androidx.fragment.app.Fragment;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.callor.todo.databinding.FragmentFirstBinding;

/**
 * findViewById 에 비해 viewBinding 방식을 사용하는 장점
 * 1. Null PointException 에서 비교적 안전하다
 * 2. xml 을 직접핸들링 하지 않기 때문에 시스템에서 좀더 여유롭다
 */
public class FirstFragment extends Fragment {

    private FragmentFirstBinding binding;

    public FirstFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        binding = FragmentFirstBinding.inflate(inflater,container,false);
        View view = binding.getRoot();
        return view;

    }

    /*
    viewBinding 방식으로 Fragment 를 사용할때
    반드시 onDestroyView() method 를 정의하고
    이 method 에서 binding 객체 변수를 null 초기화하여
    메모리 누수(메모리가 불필요하게 낭비되는 현상)를 방지해야 한다.
     */
    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: HelloApp/app/src/main/java/com/callor/helloapp/FirstFragment.kt
package com.callor.helloapp

import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.navigation.fragment.findNavController
import com.callor.helloapp.databinding.FragmentFirstBinding

/**
 * A simple [Fragment] subclass as the default destination in the navigation.
 */
class FirstFragment : Fragment() {

    /*
    코틀린의 객체 변수들은 원칙적으로 null 값 저장이 불가하다
    변수 type 뒤에 물음표를 붙여주면 제한적으로 null 값을
    저장할 수 있다
     */
    private var _binding: FragmentFirstBinding? = null

    // This property is only valid between onCreateView and
    // onDestroyView.
    private val binding get() = _binding!!

    /*
    fragment 를 확장하여 객체로 만드는 함수
    return type View?
    3개의 매개변수를 갖는 구조
     */
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {

        _binding = FragmentFirstBinding.inflate(inflater, container, false)
        return binding.root

    }

    /*
    fragment 를 화면에 그리는 함수
     */
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        /*
        firstfragment 에 있는 buttonFirst 를 클릭하면
        어딘가에 설정된 nav controller 를 찾아서
        navigate() 함수를 실행하여 화면을 전환하라
         */
        binding.buttonFirst.setOnClickListener {
            findNavController().navigate(R.id.action_FirstFragment_to_SecondFragment)
        }
    }

    /*
    화면을 지우는 함수
    화면을 그리는데 사용했던 _binding 객체를
    메모리에서 제거하는 방법
     */
    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}